### PR TITLE
Issue #932 - Does jcabi-github support Android?

### PR DESCRIPTION
**Pull request by pecko**
The ticket: https://github.com/jcabi/jcabi-github/issues/932

The PR solves two issues that disable using jcabi-github on Android or similar environment:

1) The RtGithub class creates the USER_AGENT constant using the manifest file.
Android repacks all libraries and the 'JCabi-Version' property and other similar properties doesn't  exist in apk.
All RtGithub constructors don't work. They throw:

```
java.lang.IllegalArgumentException: Attribute 'JCabi-Version' not found in MANIFEST.MF file(s) among 2 other attribute(s): ["Created-By", "Manifest-Version"]
```

Solution: 
- I have created the jcabi.properties resource file that will contain this property. The maven build updates this property (properties)

2) The second problem is described above that the Dalvik (Android JRE) can't find javax.xml.bind.DatatypeConverter
The problem happens when the issue #1 is fixed and it is related only to "new RtGithub(user, pwd)" constructor.

Solution:
- I have created the RtDatatypeConverter class based on DatatypeConverterImpl from OpenJDK.
  The class will use the standard DatatypeConverter class if exists or the method(s) copied from DatatypeConverterImpl.

I have tested the following code:

```
RtGithub github = new RtGithub("pecko", "MYPWD");
//RtGithub github = new RtGithub(oauth);
Repo repo = github.repos().get(new Coordinates.Simple("jcabi/jcabi-github"));
Issues issues = repo.issues();
```

within a Android activity.

### CODE REVIEW

**Comment by dmarkov** [comment]
@pecko Many thanks, I will find someone to review it before we merge

**Comment by dmarkov** [comment]
@darkled could you please review this PR

File: pom.xml [hunk 1]
```diff
@@ -89,10 +89,10 @@
             <artifactId>xembly</artifactId>
             <version>0.21</version>
         </dependency>
-        <dependency>
-            <groupId>com.jcabi</groupId>
-            <artifactId>jcabi-manifests</artifactId>
-        </dependency>
+<!--         <dependency> -->
```
Comment (darkled) [inline pom.xml]: @pecko do we really need to have this comments here? would it be better to remove these lines completely?

File: src/main/java/com/jcabi/github/RtDatatypeConverter.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.jcabi.github;
+
+/**
+ *
```
Comment (darkled) [inline src/main/java/com/jcabi/github/RtDatatypeConverter.java]: @pecko empty line is redundant here

File: pom.xml [hunk 3]
```diff
@@ -287,6 +287,7 @@
                         <configuration>
                             <excludes>
                                 <exclude>checkstyle:/src/test/java/com/jcabi/github/NullabilityTest.java</exclude>
+                                <exclude>checkstyle:/src/main/java/com/jcabi/github/RtDatatypeConverter.java</exclude>
```
Comment (darkled) [inline pom.xml]: @pecko let's add a puzzle to re-design this class and remove it from exclusions.

File: src/main/java/com/jcabi/github/RtGithub.java [hunk 4]
```diff
@@ -60,7 +58,7 @@
  *
  * <p>It is strongly recommended to use
  * {@link com.jcabi.http.wire.RetryWire} to avoid
- * accidental I/O exceptions:
+ * accidental I/O exceptions:import com.jcabi.manifests.Manifests;
```
Comment (darkled) [inline src/main/java/com/jcabi/github/RtGithub.java]: @pecko why do we need this import added here?

File: src/main/java/com/jcabi/github/RtGithub.java [hunk 5]
```diff
@@ -77,15 +75,13 @@
 @SuppressWarnings("PMD.TooManyMethods")
 public final class RtGithub implements Github {
 
+    // FIXME We would probably read this property from
+    // some resource property file
```
Comment (darkled) [inline src/main/java/com/jcabi/github/RtGithub.java]: @pecko let's add a puzzle here

**Comment by darkled** [comment]
@pecko please take a look at the few comments above. Also please add the original issue id and what have you done as described [here](http://www.teamed.io/qa.html):
> - Pull request description explains the solution proposed and contains a link to the original ticket it is related to.

File: pom.xml (hunk 1 above)
Comment (pecko) [inline pom.xml]: Done.

File: pom.xml (hunk 3 above)
Comment (pecko) [inline pom.xml]: Done

File: src/main/java/com/jcabi/github/RtGithub.java (hunk 4 above)
Comment (pecko) [inline src/main/java/com/jcabi/github/RtGithub.java]: Mistake. Fixed.

File: pom.xml (hunk 1 above)
Comment (darkled) [inline pom.xml]: @pecko I don't see any fixing commit from you

Comment (pecko) [inline pom.xml]: I fixed it locally.
I will commit when solving all issues.
Fighting with RtDatatypeConverter checkstyle :(

File: src/main/java/com/jcabi/github/RtDatatypeConverter.java [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.jcabi.github;
+
+/**
+ *
```
Comment (darkled) [inline src/main/java/com/jcabi/github/RtDatatypeConverter.java]: @pecko could you please also remove redundant line here?

File: src/main/java/com/jcabi/github/RtDatatypeConverter.java [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+ * The code is based on the RtDatatypeConverter from OpenJDK 7u40-b43.
+ * See https://github.com/jcabi/jcabi-github/issues/932
+ *
+ * @author Haris Peco (snpe@gmail60.com)
+ * @version $Id$
+ *
```
Comment (darkled) [inline src/main/java/com/jcabi/github/RtDatatypeConverter.java]: @pecko could you please also remove redundant line here?

File: src/main/java/com/jcabi/github/RtDatatypeConverter.java [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+ */
+public final class RtDatatypeConverter {
+
+    /**
+     * Buffer.
+     */
+    private static final int BUFFER = 64;
+        /**
```
Comment (darkled) [inline src/main/java/com/jcabi/github/RtDatatypeConverter.java]: @pecko could you please fix indentation?

File: pom.xml [hunk 9]
```diff
@@ -65,6 +65,8 @@
         <jersey.version>1.18.1</jersey.version>
         <httpcomponents.version>4.3</httpcomponents.version>
         <saxon.version>9.1.0.8</saxon.version>
+        <timestamp>${maven.build.timestamp}</timestamp>
+           <maven.build.timestamp.format>yyyy-MM-dd HH:mm</maven.build.timestamp.format>
```
Comment (darkled) [inline pom.xml]: @pecko it seems there is something wrong with the indentation here

**Comment by darkled** [comment]
@pecko please see few comments above

File: src/main/java/com/jcabi/github/RtDatatypeConverter.java (hunk 6 above)
Comment (pecko) [inline src/main/java/com/jcabi/github/RtDatatypeConverter.java]: Done.

File: src/main/java/com/jcabi/github/RtDatatypeConverter.java (hunk 7 above)
Comment (pecko) [inline src/main/java/com/jcabi/github/RtDatatypeConverter.java]: Done.

File: src/main/java/com/jcabi/github/RtDatatypeConverter.java (hunk 8 above)
Comment (pecko) [inline src/main/java/com/jcabi/github/RtDatatypeConverter.java]: Done.

File: src/main/java/com/jcabi/github/RtDatatypeConverter.java [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.jcabi.github;
+/**
+ *
```
Comment (darkled) [inline src/main/java/com/jcabi/github/RtDatatypeConverter.java]: redundant empty lines 32 and 37 still exist...

**Comment by pecko** [comment]
@darkled 
I think that I fixed all issues. The AppVeyor and Travis passed.

BTW
Travis creates builds with 3 JDK and some of that fails sometimes without obvious reason. Do you know something more about that?

**Comment by darkled** [comment]
@pecko thank you! If you find issues like this you can report a new defect and someone will try to fix it.

**Comment by darkled** [comment]
@rultor merge pls

**Comment by pecko** [comment]
@darkled 

> @pecko thank you! If you find issues like this you can report a new defect and someone will try to fix it

https://github.com/jcabi/jcabi-github/issues/1165

**Comment by yegor256** [comment]
@rultor merge pls

**Comment by pecko** [comment]
@yegor256 

There are one error and two failures in the log.

The error is weird. What OS, Maven, JDK is on rultor?
 (in com.jcabi.github.RtIssueITCase talksInGithubProject(com.jcabi.github.RtIssueITCase):
```
Tests run: 8, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.552 sec <<< FAILURE! - in com.jcabi.github.RtIssueITCase talksInGithubProject(com.jcabi.github.RtIssueITCase) Time elapsed: 0.514 sec <<< ERROR! java.lang.ClassCastException: javax.json.JsonValue$1 cannot be cast to javax.json.JsonString at org.glassfish.json.JsonObjectBuilderImpl$JsonObjectImpl.getJsonString(JsonObjectBuilderImpl.java:194) at org.glassfish.json.JsonObjectBuilderImpl$JsonObjectImpl.getString(JsonObjectBuilderImpl.java:199) at com.jcabi.github.User$Smart.name_aroundBody6(User.java:191) at com.jcabi.github.User$Smart$AjcClosure7.run(User.java:1) at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149) at com.jcabi.aspects.aj.MethodLogger.wrap(MethodLogger.java:207) at com.jcabi.aspects.aj.MethodLogger.ajc$inlineAccessMethod$com_jcabi_aspects_aj_MethodLogger$com_jcabi_aspects_aj_MethodLogger$wrap(MethodLogger.java:1) at com.jcabi.aspects.aj.MethodLogger.wrapClass(MethodLogger.java:136) at com.jcabi.github.User$Smart.name(User.java:181) at com.jcabi.github.RtIssueITCase.talksInGithubProject(RtIssueITCase.java:107)
```

My local machine, Travis and AppVeyor doesn't return this error.

The failures are here:
```
Failed tests: RtContentsITCase.canUpdateFileContentInSpecificBranch:155 Expected: "Updated" but: was "Initial." RtContentsITCase.canUpdateFileContent:110 Expected: "new content" but: was "init content"  Tests in error: RtIssueITCase.talksInGithubProject:107 » ClassCast javax.json.JsonValue$1 cann...  Tests run: 116, Failures: 2, Errors: 1, Skipped: 5
```

**Comment by yegor256** [comment]
@pecko feel free to report this as a new ticket/bug and we'll resolve it

**Comment by pecko** [comment]
@yegor256 

I have created https://github.com/jcabi/jcabi-github/issues/1169.
Just curious: Could you start rultor again? I would like to know, can the problem be reproduced always?

**Comment by yegor256** [comment]
@rultor try to merge again

**Comment by darkled** [comment]
@pecko let's ignore this test with `@Ignore` and try to merge once more

**Comment by pecko** [comment]
@darkled 

I have updated the PR (new commit: https://github.com/pecko/jcabi-github/commit/53ecd4e93afb1c6dfd0cbb752eb30acac9fcbdc5)

**Comment by darkled** [comment]
@rultor merge pls

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by pecko** [comment]
@yegor256 

Rultor fails with this error:

```
/var/lib/gems/1.9.1/gems/pdd-0.15/lib/pdd/source.rb:144:in `rescue in puzzles': ["in /home/r/repo/src/main/java/com/jcabi/github/Notification.java", #<PDD::Error: Suspicious TODO in line #41>] (PDD::Error)
```

The src/main/java/com/jcabi/github/Notification.java has not changed since January 26 - https://github.com/jcabi/jcabi-github/pull/985 (commit https://github.com/jcabi/jcabi-github/commit/c3bac3a3e0c5e8a972bc8f61ffff4e6e719081bf)

Maybe there are some changes in rultor. I think that we should try merge again.

**Comment by darkled** [comment]
@pecko let's better remove incorrect commit as described [here](http://stackoverflow.com/questions/448919/how-can-i-remove-a-commit-on-github) instead of reverting it to avoid merge conflicts in future.

**Comment by pecko** [comment]
@darkled 

> let's better remove incorrect commit as described here instead of reverting it to avoid merge conflicts in future.

I have done it. I thought that we should avoid forced push.

**Comment by darkled** [comment]
@pecko it's not a problem since you're working only with your own commits

**Comment by darkled** [comment]
@rultor try to merge pls

**Comment by pecko** [comment]
@darkled 

> it's not a problem since you're working only with your own commits

OK. Thanks.

**Comment by yegor256** [comment]
@rultor try to merge again

**Comment by darkled** [comment]
@pecko a test is failing:
```
canCreateRelease(com.jcabi.github.RtReleasesITCase) Time elapsed: 0.152 sec <<< FAILURE! java.lang.AssertionError: HTTP response status is not equal to 201
```

**Comment by pecko** [comment]
@darkled 

Why it happens? I don't touch this test (it have changed at May 21). travis (3 builds), appveyor, my local machine passed.

Also, rultor returns two different failures first time when we merge this PR.

I think that we should repeat rultor merge.

**Comment by darkled** [comment]
@yegor256 can we try once more please?

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by pecko** [comment]
This way, there is one failure:

```
{"message":"Validation Failed","errors":[{"resource":"Release","code":"custom","message":"Published releases must have a valid tag"}],"documentation_url":"https://developer.github.com/v3/repos/releases/#create-a-release"}>) out of org.hamcrest.MatcherAssert#assertThat[20] in 144ms Tests run: 6, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 4.679 sec <<< FAILURE! - in com.jcabi.github.RtReleaseAssetsITCase uploadsSameAssetInTwoReleases(com.jcabi.github.RtReleaseAssetsITCase) Time elapsed: 0.161 sec	<<< FAILURE!
```

We should probably open a new issue and investigate that. 

Is it possible run rultor test merge (or build) locally?

**Comment by yegor256** [comment]
@pecko indeed, don't hesitate to create a new ticket for this problem. You can run what rultor is doing, but locally. Just install Docker and then create a new container from `yegor256/rultor` image. Inside the container run `mvn clean install`

**Comment by pecko** [comment]
@yegor256 

I think that I have found why it happens.
When we run test locally we don't run integration tests. The RtReleaseAssetsITCase test doesn't run with

```
mvn clean install -Pqulice
```

that is, in our local tests, Travis, AppVeyor ...

Rultor run this test and fails. The issue is that it fails randomly (we have had 0 failures, 2 failures, 1 failures ..). I'm not sure if it is rultor or jcabi-github issue (probably jcabi-github)

I can try to run this test locally, but can I use
```
mvn clean install -Dit.test=RtReleaseAssetsITCase -Dfailsafe.github.key=<token> -Dfailsafe.github.key.second=<second-token> -Dfailsafe.github.repo=<repo>
```

**Comment by yegor256** [comment]
@pecko you can find this information at the bottom of `README.md`

**Comment by pecko** [comment]
@yegor256 

I have tried the following:

```
mvn  -Dit.test=RtReleaseAssetsITCase -Dfailsafe.github.key=<my_key> test
mvn  -Dit.test=RtReleaseAssetsITCase -Dfailsafe.github.key=<my_key> verify
mvn  clean install -Dit.test=RtReleaseAssetsITCase -Dfailsafe.github.key=<my_key>
```
(I have generated API key on github with all required scopes)

run only this test as JUnit test from Eclipse.
Eclipse calls setUp/tearDown, but no one method is called.

Travis and AppVeyor run it.

I will investigate still

**Comment by pecko** [comment]
@yegor256 

Please forget my previous comment. Tests run correctly, but some integration tests fail randomly.
I have create https://github.com/jcabi/jcabi-github/issues/1170

**Comment by pecko** [comment]
Please see https://github.com/jcabi/jcabi-github/issues/1170#issuecomment-143884826
Could you try merge rultor again?
These tests have passed on rultor once (we have had the issue with Notification todo in that case).

**Comment by yegor256** [comment]
@pecko don't forget to address your comments. Otherwise, how do we know who are you talking to?

**Comment by yegor256** [comment]
@rultor try to merge again

**Comment by pecko** [comment]
@yegor256 

Now, we have two failures in a new class (RtContentsITCase):

```
Tests run: 8, Failures: 2, Errors: 0, Skipped: 1, Time elapsed: 7.309 sec <<< 
FAILURE! - in com.jcabi.github.RtContentsITCase getContent(com.jcabi.github.RtContentsITCase) Time elapsed: 1.032 sec <<< FAILURE!
````

#1170 requires more investigation.

**Comment by pecko** [comment]
@yegor256 

> don't forget to address your comments. Otherwise, how do we know who are you talking to?

Sorry again. Have to exercise. 

This is weird issue. Does it happen before with this project?

**Comment by yegor256** [comment]
@pecko "unstable build" is a very common problem for projects that have integration tests. The only solution is to create a new ticket every time you see a problem and fix it.

**Comment by pecko** [comment]
@yegor256 

OK. I understand. I have already created #1170.
What else we can do with this ticket?

**Comment by darkled** [comment]
@pecko I would suggest to ignore failing tests and merge PR, then ask to un-ignore these tests in #1170.

**Comment by pecko** [comment]
@darkled 

I will try. I hope that some other won't fail.

**Comment by pecko** [comment]
@darkled

The https://github.com/jcabi/jcabi-github/commit/0ba5c7768012e19c56fe8aaae36d8e207f995ce9 ignores three tests mentioned in #1170

**Comment by darkled** [comment]
@yegor256 try to merge once more please

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by pecko** [comment]
@darkled  @yegor256 

A new error

```
/home/r/repo/src/test/java/com/jcabi/github/RtNotificationsITestCase.java", #<PDD::Error: Suspicious TODO in line #36>] (PDD::Error)
```

I will try to fix it.

**Comment by pecko** [comment]
@darkled @yegor256

I have found a few invalid todos and fix them. The commit: https://github.com/jcabi/jcabi-github/commit/bd2e007706e610b52b89d7e30277e5ed0fcf8c76

**Comment by pecko** [comment]
@darkled @yegor256

I'm sorry because repeat message, but could you please try to merge this PR once yet? There was a few invalid todos that are fixed.

**Comment by darkled** [comment]
@rultor merge once more please

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by pecko** [comment]
@darkled  @yegor256 

Rultor fails in new test:

```
Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.354 sec 
<<< FAILURE! - in com.jcabi.github.RtReposITCase create(com.jcabi.github.RtReposITCase) Time elapsed: 1.111 sec <<< FAILURE!
```

Will we ignore this test too?

**Comment by darkled** [comment]
@pecko I am not sure, it seems we have problems with integration tests and cannot merge anything before these problems fixed.
@yegor256 are we waiting till #1170 is fixed?

**Comment by yegor256** [comment]
@darkled yes, seems so

**Comment by pecko** [comment]
@yegor256 @darkled 

How it worked before? This PR introduce nothing related to these failures, errors, todos ...

**Comment by darkled** [comment]
@pecko well, it worked pretty fine before. as you can see in the closed puzzles. I think the problem came with some of previous merges, since other PRs has the same failures.

**Comment by pecko** [comment]
@darkled 

I want just to understand it better. You claim that integration tests never fail before or they fail rarely?

**Comment by darkled** [comment]
@pecko everything what happened before is on github, so you can easily check previous PRs :) I didn't see anything similar to our case. But now there are few PRs with very similar problems.

**Comment by pecko** [comment]
@darkled 

I believe you :) . If they haven't fails before, we have some weird issue that can be in different projects (rultor, jcabi-github, jcabi-http...). I hope, we will fix it soon.

**Comment by amihaiemil** [comment]
@pecko This PR is too big and too old, anyway. There is another PR for removing the Manifests mechanisms, so that issue will be solved there. If you care to make another PR with smaller changes, that would improve the Android issue, it would be welcome.

**Comment by 0crat** [comment]
The job is not in WBS, won't close the order

**Pull request closed without merging**